File: flutter/lib/domain/entities/cleaned_url.dart
import 'package:equatable/equatable.dart';
import 'url_parameter.dart';

class CleanedUrl extends Equatable {
  final String originalUrl;
  final String baseUrl;
  final String cleanedUrl;
  final List<UrlParameter> parameters;
  final int charactersSaved;
  final DateTime cleanedAt;

  const CleanedUrl({
    required this.originalUrl,
    required this.baseUrl,
    required this.cleanedUrl,
    required this.parameters,
    required this.charactersSaved,
    required this.cleanedAt,
  });

  factory CleanedUrl.fromUrl(String url, List<UrlParameter> parameters) {
    final selectedParams = parameters.where((p) => p.selected).toList();
    final queryString = selectedParams.map((p) => p.toQueryString()).join('&');
    final baseUrl = _extractBaseUrl(url);
    final cleanedUrl = queryString.isEmpty ? baseUrl : '$baseUrl?$queryString';
    
    return CleanedUrl(
      originalUrl: url,
      baseUrl: baseUrl,
      cleanedUrl: cleanedUrl,
      parameters: parameters,
      charactersSaved: url.length - cleanedUrl.length,
      cleanedAt: DateTime.now(),
    );
  }

  static String _extractBaseUrl(String url) {
    final uri = Uri.tryParse(url);
    if (uri == null) return url;
    return '${uri.origin}${uri.path}';
  }

  int get totalParameters => parameters.length;
  int get selectedParameters => parameters.where((p) => p.selected).length;
  int get trackerCount => parameters.where((p) => p.isTracker).length;
  double get compressionRatio => charactersSaved / originalUrl.length;

  @override
  List<Object?> get props => [
        originalUrl,
        baseUrl,
        cleanedUrl,
        parameters,
        charactersSaved,
        cleanedAt,
      ];

  @override
  bool get stringify => true;
}

File: react_native/src/domain/entities/CleanedUrl.ts
import { UrlParameter } from './UrlParameter';

export interface CleanedUrl {
  originalUrl: string;
  baseUrl: string;
  cleanedUrl: string;
  parameters: UrlParameter[];
  charactersSaved: number;
  cleanedAt: Date;
}

export const createCleanedUrl = (
  originalUrl: string,
  parameters: UrlParameter[]
): CleanedUrl => {
  const url = new URL(originalUrl);
  const baseUrl = `${url.origin}${url.pathname}`;
  
  const selectedParams = parameters.filter(p => p.selected);
  const queryString = selectedParams
    .map(p => `${p.key}=${p.value}`)
    .join('&');
  
  const cleanedUrl = queryString ? `${baseUrl}?${queryString}` : baseUrl;
  const charactersSaved = originalUrl.length - cleanedUrl.length;
  
  return {
    originalUrl,
    baseUrl,
    cleanedUrl,
    parameters,
    charactersSaved,
    cleanedAt: new Date(),
  };
};

export const getUrlStatistics = (cleanedUrl: CleanedUrl) => ({
  totalParameters: cleanedUrl.parameters.length,
  selectedParameters: cleanedUrl.parameters.filter(p => p.selected).length,
  trackerCount: cleanedUrl.parameters.filter(p => p.isTracker).length,
  compressionRatio: cleanedUrl.charactersSaved / cleanedUrl.originalUrl.length,
});